pwndbg/gdblib/disasm/arm.py

```diff
@@ -30,10 +30,10 @@ def _condition(self, instruction: PwndbgInstruction, emu: Emulator) -> Instructi
             else pwndbg.gdblib.regs.xpsr
         )
 
-        N = value & (1 << 31)
-        Z = value & (1 << 30)
-        C = value & (1 << 29)
-        V = value & (1 << 28)
+        N = (value >> 31) & 1
+        Z = (value >> 30) & 1
+        C = (value >> 29) & 1
+        V = (value >> 28) & 1
```

Comment:
I'm okay with those changes but we can as well skip them?

Not really that important, but the `1<<31` etc. is optimized here by Python's peephole optimizer while in `(value >> 31) & 1` all the operations are computed.

This can be seen here:
<img width="1059" alt="image" src="https://github.com/pwndbg/pwndbg/assets/10009354/50c29b5b-56f7-405b-b122-75e23bee8444">

Note that they also added specialized instructions/bytecode in Python 3.11 which means that some hot used functions may be optimized/specialized for the very-often used argument types. This is also why I am providing the `show_caches=1, adaptive=1` keyword arguments here. This may also influence the benchmark result here, but oh well, its worth using the latest tech ;). Fwiw I have given a talk on the [Python specialized bytecode here](https://www.youtube.com/watch?v=RlNM5n5C_wg&t=5034s).

Reply:
In this particular case, the functionality was incorrect with the old implementation.

There were some checks for equality between the different bit flags, and the old way of extracting the bit didn't make it 0/1, but kept the bit values in their bit positions, so some equality checks didn't pass when they should.

https://github.com/pwndbg/pwndbg/blob/5c84061262055cedb1d4b4233afeba80e7144045/pwndbg/gdblib/disasm/arm.py#L49-L52

The Python opcode optimization is very interesting, I'll have to dig deeper into that.